New simulation model

GernotMaier opened the pull request:
This is a very large pull request with several major changes:

- changing names of telescopes to official CTAO naming (e.g., MSTN-01). This required a re-writing of the model database and many changes in the code
    - [ ] open point: update all docstrings and documentation for new telescope naming
-  new organisation of model parameters with a base `model_parameter` class and to specialized classes called `telescope_model` and `site_model` (allows to add in future e.g., illuminator_model classes)
- reading of simulation model parameters from json files (either from file from from repository)
   - directory / repository structure outlined in [simulation_model README](https://github.com/gammasim/simulation_model/blob/main/README.md)
- all site and telescope-related parameters are now in the database (or in the simulation model repository)
  - this affects especially the data in the [data/parameters/corsika_parameters.yml](data/parameters/corsika_parameters.yml) file and the auxiliary data in the metadata headers of the array element / telescope lists. 
-  additional changes for more consistent naming (e.g., in db instead of `units` use `unit` (as in astropy)

Given that telescope naming has changed, this PR required a lot of name-but-not-functional changes in code, tests, and data files used.

Temporary items to be fixed best in a future PR:

- simulation_model is expected to be on GitHub (planned to be moved to CTAO gitlab)
- CIs for unit- and integration tests required the cloning of the simulation model (see [here](https://github.com/gammasim/simtools/blob/c54be3754520013ca89cf0a9c0929b557029bcb4/.github/workflows/CI-integrationtests.yml#L46) and [here](https://github.com/gammasim/simtools/blob/c54be3754520013ca89cf0a9c0929b557029bcb4/.github/workflows/CI-unittests.yml#L101)). This should go when all parameters are in the simulation model database
- old/new names are currently duplicated in names.py (see [here](https://github.com/gammasim/simtools/blob/c54be3754520013ca89cf0a9c0929b557029bcb4/simtools/utils/names.py#L121). This also requires changes in the database.
- missing documentation updates (this is probably a more general issue)

Suggestion for review:

- a lot of files changed. 
   - suggest first to do the data files in tests/resources: changes are in telescope names (e.g., from LST-1 to LSTN-01) and removal of corsika-related parameters from the metadata header (now included in db_handler)
   - data/parameters/corsika_parameters.yml : all site/telescope related parameters are no in the DB
- simple changes: `db_handler` is now in `db/db_handler` (as there are more db-related modules). This obviously changes all imports on db_handler
- for actual code review: best maybe to start with `names.py`, then `db_handler` and then the model parameter classes (`model_parameter`, `telescope_model`, `site_model`). 
- after that, look at the rest of the modules and tests (lots of little and annoying changes)

GernotMaier commented:
Nice progress before the weekend: I can now run 

` python simtools/applications/print_array_elements.py --config tests/integration_tests/config/print_array_elements_utm_to_ground_meta_in_table_par_from_db.yml`

with print_array_elements_utm_to_ground_meta_in_table_par_from_db being:
```
CTA_SIMPIPE:
  APPLICATION: simtools-print-array-elements
  TEST_NAME: utm_to_ground_meta_in_table_par_from_db
  CONFIGURATION:
    INPUT: tests/resources/telescope_positions-North-utm-no-corsika-par.ecsv
    EXPORT: ground
    USE_CORSIKA_TELESCOPE_HEIGHT: true
    OUTPUT_PATH: simtools-tests
    OUTPUT_FILE: test.ecsv
    SELECT_ASSETS: ['LST', 'MST']
    LOG_LEVEL: DEBUG
  INTEGRATION_TESTS:
    - REFERENCE_OUTPUT_FILE: |
        ./tests/resources/telescope_positions-North-ground.ecsv
      TOLERANCE: 1.e-2
```

with reading all relevant parameters from the database / simulation model repositories. This includes site parameter (e.g., altitude), and the corsika sphere radius / center values.

Most tests are still broken and a lot of `TODO`s all over the place.

GernotMaier commented on .env_template:
`SIMTOOLS_DB_SIMULATION_MODEL_URL` allows to read additional model parameters (or update existing ones) from a file system or repository (e.g., this one https://github.com/gammasim/simulation_model)

GernotMaier commented on .github/workflows/CI-integrationtests.yml:
The new simulation model parameters for site and telescopes (sphere radius, axis height) are not yet in the database. This is why we need to clone the simulation model repository and use it for integration tests.

GernotMaier commented on .github/workflows/CI-unittests.yml:
The new simulation model parameters for site and telescopes (sphere radius, axis height) are not yet in the database. This is why we need to clone the simulation model repository and use it for unit tests.

GernotMaier commented on data/layout/telescope_positions-North-1LST.ecsv:
This and all following telescope position files: all these data moved into the database and can therefore be removed from the headers.

GernotMaier commented on tests/integration_tests/config/compare_cumulative_psf_prod5_psf.yml:
This and following config files: most changes are for the new telescope naming.

GernotMaier commented on tests/resources/prod_config_test.yml:
One thing to note now is that there is no FlashCam in the North anymore.

GernotMaier commented on tests/resources/reference_point_altitude.json:
These are the changes applied to keys in the DB (see also the code). See #799

GernotMaier commented on simtools/applications/add_file_to_db.py:
`db_handler` moved to `simtools/db/db_handler`. So this change you will see a lot.

GernotMaier commented on simtools/data_model/metadata_collector.py:
These changes are introduced to get easily the site from the metadata. This was written before we agreed that we always want to have a `site` field in the main DB fields. So for now, it is not directly used, but I think still useful and I would like to keep it into the code.

GernotMaier commented on simtools/layout/array_layout.py:
This modules includes a lot of changes: no reading of the headers for the site and corsika parameters, but query the DB to retrieve them.

GernotMaier commented on simtools/model/telescope_model.py:
A lot of the code is moved to the generic simtools/model/model_parameter module.

VictorBarbosaMartins reviewed (changes requested):
Hi @GernotMaier. 

Given the very long PR, this is also a very long review. Two main points have been discussed in the group but I summarize them here:

- Unit tests: unit tests are not complete but will be completed in a next step. They are also very slow but that will also be dealt with in one of the next PRs;

- The implementation of the ModelParameter (which should actually be ParameterModel): as discussed, this could be improved by defining a child class to assign for each parameter. This would standardise the methods and attributes of all parameters. I would suggest checking at least if this is an option for you.

In addition to that, I have found today that the tools were broken by some of these changes. I also commented about this in the comments below but at that time I thought the error was in one tool. Now I see that they are in many places. The same error is appearing in the docker checks in github, so this is likely something very fundamental. Test it with your favorite application by running e.g. `simtools-print-array-elements --config tests/integration_tests/config/print_array_elements_ground_to_ground.yml`. Interestingly the github integration tests are passing (not sure why running by hand does not work).

Apart from that, please ignore my comments in regard to missing tests. I have written those comments before we spoke about it. I pushed only a couple of typos/grammar that I found.

I hope you find at least some of the comments useful. We can discuss about them in person tomorrow. Thanks!

GernotMaier commented:
@VictorBarbosaMartins - thanks a lot for the review! Clearly that was a tough one.

I will go through your comments and address each of them. As discussed:

- will close the unit test related one
- some of your comments relate to quite old code which has been moved but not changed; I'll see what to do there

For the application which is not working for you locally: not all simulation model parameters are in the DB yet and this PR still needs a local copy of the simulation model repository.

e.g.
```
git clone https://gitlab.cta-observatory.org/cta-science/simulations/simulation-model/model_parameters.git
```
then set in your `.env` file:
```
SIMTOOLS_DB_SIMULATION_MODEL_URL="../model_parameters/"
```
(or wherever you cloned the repo above to).

Please test the application again.

GernotMaier commented on tests/resources/array_config_test.yml:
```diff
@@ -2,9 +2,9 @@ site: North
-MST-05:
-    name: NectarCam-D
+    LSTN: 'design'
+    MSTN: 'design'
```
This has been extensively discussed, e.g., #799 .

Design is not the latest implementation, but it is the placeholder for the nominal model used for the time being. At some point, we will have a model for each telescope.

GernotMaier commented on tests/resources/reference_position_mercator.schema.yml:
This is replaced by tests/resources/reference_point_altitude.json - the submit data test for this is not required (we do the `validate_file_using_schema_json..` tests, which covers this.

GernotMaier commented on tests/unit_tests/db/test_db_handler.py:
```diff
@@ -295,48 +300,63 @@ def test_update_parameter_field_db(db, random_id, db_cleanup, io_handler):
+            assert _obs_level == pytest.approx(2158.0)
     else:
-        assert pars["altitude"] == 2158
+        assert pars["altitude"] == 2156
```
It changes as we use now the prod6 value.

GernotMaier commented on tests/unit_tests/db/test_db_handler.py:
```diff
@@ -295,48 +300,63 @@ def test_update_parameter_field_db(db, random_id, db_cleanup, io_handler):
     pars = db.get_site_parameters("South", "Released")
     if db.mongo_db_config:
-        assert pars["altitude"]["Value"] == 2147
+        # temporary solution for simulation model parameter renaming
+        if "corsika_observation_level" in pars:
+            _obs_level = pars["corsika_observation_level"].get("value") or pars[
+                "corsika_observation_level"
+            ].get("value")
+        else:
+            _obs_level = pars["altitude"].get("value") or pars["altitude"].get("value")
+        assert _obs_level == pytest.approx(2147.0)
```
Fixed.

GernotMaier commented on tests/unit_tests/db/test_db_from_repo_handler.py:
```diff
@@ -0,0 +1,123 @@
+        model_version=None,
+        db_simulation_model_url=db_config["db_simulation_model_url"],
+    )
+    assert len(_pars) > 0
```
I understand this - but the issue is that we hardcode model values into the tests. I know we do this in other places, but I don't think this testing the functionality (we have otherwise discussions like on the previous item where we wonder why the height changes by 2 m).

In this case we want to be sure that parameters are read from the repository, and a non-zero length test on the return dictionary is covering this sufficiently?

VictorBarbosaMartins replied:
it might be that all values are Nan, or 0 if something is broken, but the length would be still correct. we can leave it for another PR when dealing only with the tests.

GernotMaier commented on tests/integration_tests/config/print_array_elements_print_compact.yml:
- `compact` command line has been replaced by a clearer `print` command
- some functionality is now tested with `tests/integration_tests/config/print_array_elements_print_ground.yml`

GernotMaier commented on tests/integration_tests/config/plot_array_layout_layout_name_one_angle.yml:
```diff
@@ -2,6 +2,7 @@ CTA_SIMPIPE:
   TEST_NAME: layout_name_one_angle
   CONFIGURATION:
+    SITE: North
     ARRAY_LAYOUT: North-4LST
```
We are using it here to find easier the layout file `data/layout/telescope_positions-North-4LST.ecsv`. I agree it is duplicated, but it simplifies the tests here.

But you are right that there is some duplication on some cases. I've opened in issue to address this: #816

GernotMaier commented on tests/unit_tests/db/test_db_from_repo_handler.py:
```diff
@@ -0,0 +1,123 @@
+
+    _pars_telescope_model = ["telescope_axis_height", "telescope_sphere_radius"]
+
+    _pars_mstn01 = db_from_repo_handler._update_parameters_from_repo(
+        parameters=dict.fromkeys(_pars_telescope_model, None),
+        site="North",
+        telescope_name="MSTN-01",
+        model_version=None,
+        parameter_collection="telescope",
+        db_simulation_model_url=db_config["db_simulation_model_url"],
+        db_simulation_model="verified_model",
+    )
+    assert len(_pars_mstn01) > 0
+    assert "telescope_axis_height" in _pars_mstn01
+    for key in ["value", "unit", "site"]:
+        assert key in _pars_mstn01["telescope_axis_height"]
+
+    _pars_mstndesign = db_from_repo_handler._update_parameters_from_repo(
+        parameters=dict.fromkeys(_pars_telescope_model, None),
+        site="North",
+        telescope_name="MSTN-design",
+        model_version=None,
+        parameter_collection="telescope",
+        db_simulation_model_url=db_config["db_simulation_model_url"],
+        db_simulation_model="verified_model",
+    )
+    assert len(_pars_mstndesign) > 0
```
Good point, done.

GernotMaier commented on simtools/db/db_from_repo_handler.py:
```diff
@@ -0,0 +1,194 @@
+        _design_model = None
+    else:
+        logger.error(f"Unknown parameter collection {parameter_collection}")
+        raise ValueError
```
Yes. Added.

GernotMaier commented on simtools/db/db_from_repo_handler.py:
```diff
@@ -0,0 +1,194 @@
+
+    for key in parameters.keys():
+        _parameter_file = gen.join_url_or_path(_file_path, f"{key}.json")
+        try:
+            parameters[key] = gen.collect_data_from_file_or_dict(
+                file_name=_parameter_file, in_dict=None
+            )
+        except (FileNotFoundError, gen.InvalidConfigData):
+            # use design telescope model in case there is no model defined for this telescope ID
+            # accept errors, as not all parameters are defined in the repository
+            try:
+                _parameter_file = gen.join_url_or_path(
+                    db_simulation_model_url, db_simulation_model, _design_model, f"{key}.json"
+                )
+                parameters[key] = gen.collect_data_from_file_or_dict(
+                    file_name=_parameter_file, in_dict=None
+                )
+            except (FileNotFoundError, TypeError, gen.InvalidConfigData):
+                pass
+
```
I've looked over it again and I am not sure I agree that adding a function would simplify it:

- it is a two step process, so it is dealing first with one set of error, then with (mostly the same second set of errors)
- we only know after accessing the files if the parameter is defined. So a separate function would do exactly this.

Do you have a concrete suggestion how to make it shorter?

VictorBarbosaMartins replied:
ok, then let's keep it like this and perhaps add some logger debug messages to help (just in case).

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -39,11 +39,14 @@ class DatabaseHandler:
+    #    DB_TABULATED_DATA = "CTA-Simulation-Model"
+    #    DB_CTA_SIMULATION_MODEL = "CTA-Simulation-Model"
+    # DB collection with updates field names
+    DB_TABULATED_DATA = "Staging-CTA-Simulation-Model"
+    DB_CTA_SIMULATION_MODEL = "Staging-CTA-Simulation-Model"
```
This is the general name used widely (Orel googled it):

- we apply updates to models (and therefore the DB), but won't put these updates immediately into the production DB
- this is why the Staging DB is introduced - we can do all the tests and when we are happy this is propagated to production

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -139,32 +136,32 @@ def get_model_parameters(
+                db_simulation_model_url=self.mongo_db_config.get("db_simulation_model_url", None),
+            )
+
+        return _pars
```
Done (it requires a deep_copy of the db to make sure that other tests are not affected).

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -216,110 +213,29 @@ def export_model_files(self, parameters, dest):
-                self._write_model_file_yaml(value, dest, no_file_ok=True)
+        if self.mongo_db_config.get("db_simulation_model_url", None) is not None:
+            self._logger.warning(
+                "Exporting model files from simulation model repository not yet implemented"
```
It is not implemented yet - this is about model parameters saved in files (e.g., the quantum efficiency table). The `db_from_repo_handler` module cannot handle this yet.

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -440,62 +327,39 @@ def read_mongo_db(
-            if _parameters[par_now]["File"] and write_files:
-                file = self._get_file_mongo_db(db_name, _parameters[par_now]["Value"])
+            if _parameters[par_now]["file"] and write_files:
+                file = self._get_file_mongo_db(db_name, _parameters[par_now]["value"])
```
I resolve this, as this is one of the plenty of tests needed to be added for db_handler

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -1078,9 +874,9 @@ def update_parameter_field(
+        allowed_fields = ["applicable", "unit", "type", "items", "minimum", "maximum"]
         if field not in allowed_fields:
-            raise ValueError(f"The field to change must be one of {', '.join(allowed_fields)}")
+            raise ValueError(f"The field {field} must be one of {', '.join(allowed_fields)}")
```
Same thing here: unit tests will be added as one of the next steps.

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -1089,17 +885,17 @@ def update_parameter_field(
+            query["instrument"] = telescope
+            logger_info = f"instrument {telescope}"
+        elif site is not None and site in names.site_names:
+            query["site"] = site
```
Same thing here: unit tests will be added as one of the next steps.

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -1257,25 +1053,25 @@ def add_new_parameter(
+            db_entry["instrument"] = names.validate_telescope_name(telescope)
         elif "sites" in collection_name:
-            db_entry["Site"] = names.validate_site_name(site)
+            db_entry["site"] = names.validate_site_name(site)
```
Same thing here: unit tests will be added as one of the next steps.

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -1178,18 +974,18 @@ def add_parameter(
         files_to_add_to_db = set()
         if self._is_file(new_value):
-            par_entry["File"] = True
+            par_entry["file"] = True
```
Same thing here: unit tests will be added as one of the next steps.

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -1257,25 +1053,25 @@ def add_new_parameter(
+        db_entry["file"] = False
         if self._is_file(value):
-            db_entry["File"] = True
+            db_entry["file"] = True
```
Same thing here: unit tests will be added as one of the next steps.

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -1472,9 +1265,42 @@ def get_all_available_telescopes(
 
         return _all_available_telescopes
+
+    def get_telescope_db_name(self, telescope_name):
```
I think you are right, but I see that this isn't a quick fix: the functions in names.py (e.g., validate_telescope_name and get_telescope_type_from_telescope_name) need to be renamed as well. As they are are widely used, I prefer to postpone this into an upcoming issue: #817

GernotMaier commented on simtools/data_model/metadata_collector.py:
```diff
@@ -186,12 +215,12 @@ def _fill_associated_elements_from_args(self, associated_elements_dict):
+                _telescope_name = names.validate_telescope_name(self.args_dict["telescope"])
+                _association["class"] = "telescope"
+                _association["type"] = names.get_telescope_type_from_telescope_name(_telescope_name)
+                _association["subtype"] = ""
```
`subtype` is given by the CTAO metadata schema. I would prefer to keep it therefore.

GernotMaier commented on .github/workflows/DB-Backup.yml:
```diff
@@ -0,0 +1,64 @@
+on:
+  schedule:
+    - cron: '0 2 * * *'  # Run every night at 2 am
+  workflow_dispatch:     # Manual trigger
```
I am puzzled that this appears here - the DB backup functionality has been reviewed in PR #805.

And now - this runs on a schedule or manually.

GernotMaier commented on simtools/utils/names.py:
```diff
@@ -19,59 +17,53 @@
+# Telescopes and other array elements
+array_element_names = {
+    # CTAO telescopes
+    "LSTN": {"site": "North", "observatory": "CTAO", "class": "telescope"},
+    "MSTN": {"site": "North", "observatory": "CTAO", "class": "telescope"},
+    "LSTS": {"site": "South", "observatory": "CTAO", "class": "telescope"},
+    "MSTS": {"site": "South", "observatory": "CTAO", "class": "telescope"},
+    "SSTS": {"site": "South", "observatory": "CTAO", "class": "telescope"},
+    "SCTS": {"site": "South", "observatory": "CTAO", "class": "telescope"},
```
I've tried that just now and it is much less readable. Above let's you see the differences immediately, adding helper dictionaries requires to read line by line. Hope it is ok if we leave it as it is.

GernotMaier commented on simtools/utils/names.py:
```diff
@@ -105,60 +98,59 @@
-    "scts": ["SCTS", "scts"],
+# List of site parameters which are not part of the simtel configuration
+# or which have different naming in the database and simtel configuration.
+# simtel: True if this is a simtel parameter (allows to give alternative "name")
```
It is not in all cases false - `corsika_observation_level` has it set to true. There might be more in future.

GernotMaier commented on simtools/utils/names.py:
```diff
@@ -105,60 +98,59 @@
+    "reference_point_latitude": {"db_name": "ref_lat", "simtel": False},
+    "reference_point_utm_north": {"db_name": "reference_point_utm_north", "simtel": False},
+    "reference_point_utm_east": {"db_name": "reference_point_utm_east", "simtel": False},
+    # Note naming inconsistency between old and new model
+    # altitude was the corsika observation level in the old model
```
Removed the duplicated comment.

We will need to have a mapping of model names to simtel names. In principle this can be also listed in the schema, but we don't want to access schema files each time we want a model name.

GernotMaier commented on simtools/utils/names.py:
```diff
@@ -105,60 +98,59 @@
+    "reference_point_utm_east": {"db_name": "reference_point_utm_east", "simtel": False},
+    # Note naming inconsistency between old and new model
+    # altitude was the corsika observation level in the old model
+    "corsika_observation_level": {"db_name": "altitude", "simtel": True},
```
It is also used in simtel.

GernotMaier commented on simtools/utils/names.py:
```diff
@@ -105,60 +98,59 @@
-    """
-    return _validate_name(name, all_telescope_class_names)
+# List of telescope parameters which are not part of the simtel configuration
+# or which have a different name in the simtel configuration.
```
This replaces the list of parameters not to write in top of simtel_config_writer, see [here](https://github.com/gammasim/simtools/blob/acdade26d1a6cab137fc5abf4ebffead7c4c35b2/simtools/simtel/simtel_config_writer.py#L37). I figured it out that it is better to have all of this in a single dict and need distributed in several modules.

GernotMaier commented on simtools/utils/names.py:
```diff
@@ -105,60 +98,59 @@
-
-    Valid names e.g.,
-    - D
+    Validate telescope ID. Allowed IDs are
```
An ID does not need to be a number (e.g., a UUID is a string).

GernotMaier commented on simtools/utils/names.py:
```diff
@@ -257,322 +251,199 @@ def _validate_name(name, all_names):
 
 
-def validate_telescope_name_db(name):
+def get_list_of_telescope_types(array_element_class="telescope", site=None, observatory="CTAO"):
```
Not a python class! e.g., class of array_elements which can be telescope, calibration, ...

This follow the CTAO convention outlined in the metadata document.

And again, I agree that this is one of the functions which should be addressed by #817

GernotMaier commented on simtools/applications/plot_array_layout.py:
```diff
@@ -147,6 +168,11 @@ def main():
+        site = (
+            _get_site_from_telescope_list_name(one_file)
+            if args_dict["site"] is None
+            else args_dict["site"]
```
I am not familiar enough with it - site is used at several places in the application so I've added it.

GernotMaier commented on simtools/applications/print_array_elements.py:
```diff
@@ -15,20 +15,19 @@
         File name with list of array element positions
-    compact (str)
-        Compact output in requested coordinate system; possible are corsika,utm,mercator
+    print (str)
+        Print requested coordinate system; possible are ground, utm, mercator
     export (str)
         Export array element list to file in requested coordinate system; \
```
Yes, this is why the description says "print requested coordinate system". I think adding a third parameter would make this more complicated.

(and apologies to add this unrelated change to this PR).

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+    ----------
+    site: str
+        Site name (e.g., South or North).
+    telescope_model_name: str
```
At this point we only have telescopes. But good point and I adding it to PR #817 (as this requires quite a bit of renaming and I don't want to mix it now with all the other changes)

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+            self.db = db_handler.DatabaseHandler(mongo_db_config=mongo_db_config)
+
+        self._parameters = {}
+        self._derived = None
```
`model_parameter.derived` is a derived model parameter. This has been taken over from before and it is nice and short. Do we need longer description?

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        self._parameters = {}
+        self._derived = None
+        self.site = names.validate_site_name(site) if site is not None else None
+        self.name = (
```
Could we keep that for now? It is used all over the place and we anyway want to change this with #817 into instrument_name (or something similar).

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        par_name: str
+            Name of the parameter.
+
+        Returns
+        -------
+        Value of the parameter
```
I've renamed `get_parameter` to `get_parameter_dict` and adjusted also the comments.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        ----------
+        par_name: str
+            Name of the parameter.
+        parameter_dict: dict
```
Isn't this clear from the first line of this function?

`parameter_dict = parameter_dict if parameter_dict else self.get_parameter_dict(par_name)`

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+    def get_parameter_value_with_unit(self, par_name):
+        """
+        Get the value of an existing parameter of the model as an Astropy Quantity with its unit.\
+        If no unit is provided in the model, the value is returned without a unit.
```
Yes and no. You are right that the function name is a bit misleading for parameters without a unit. But the way we need to use it, dimensionless parameter should not be a Quantity.

Is this ok or do you suggest a renaming of the function? A good suggestion for a name?

VictorBarbosaMartins replied:
ok, if this is used only in one place and that is how we need it. Although, I would tend to get the value with `.value` in the code using this function instead of directly in the return statement. anyways, we can keep it like this.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+
+        Raises
+        ------
+        InvalidModelParameter
```
Fixed.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+
+    def get_parameter_type(self, par_name):
+        """
+        Get the type of an existing parameter of the model.
```
Done.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+            type of the parameter (None if no type is defined)
+
+        """
+        parameter_dict = self.get_parameter(par_name)
```
It returns a dict, and we have renamed it earlier (see above)

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+            self._logger.debug(f"Parameter {par_name} does not have a type")
+        return None
+
+    def get_parameter_file_flag(self, par_name):
```
I think this is clear with familiarity of the DB entries: there is a 'file' flag (boolean).

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        """
+        parameter_dict = self.get_parameter(par_name)
+        try:
+            if parameter_dict.get("file") or parameter_dict.get("File"):
```
This would require a renaming of all database entries. I suggest we do not do that and keep it as it is.

I've added one sentence to the docstring saying that this is a boolean.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        """
+        if self._derived is None:
+            self._load_derived_values()
+            self.export_derived_files()
```
OK. Fixed.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        Load derived values from the DB
+
+        """
+        self._logger.debug("Reading derived data from DB")
```
I've renamed it to `values`.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+
+        """
+
+        if self.name is None:
```
See discussion earlier on self.name.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+
+    def _set_config_file_directory_and_name(self):
+        """
+        Set and create the directory model parameter files are written to.
```
Fixed.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+            )
+            self._parameters.update(_site_pars)
+
+    def set_extra_label(self, extra_label):
```
It is used in ./simtools/model/array_model.py

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+
+    def get_simtel_parameters(self, telescope_model=True, site_model=True):
+        """
+        Get simtel parameters as name and value pairs. Do not include parameters
```
Yes, but the second sentence makes it clearer where this is defined and therefore easier to find the code.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+
+        self._logger.info(f"Adding {par_name}={value} to the model")
+        self._parameters[par_name] = {}
+        self._parameters[par_name]["value"] = value
+        self._parameters[par_name]["type"] = type(value)
+        self._parameters[par_name]["applicable"] = is_applicable
+        self._parameters[par_name]["file"] = is_file
```
Need to check - didn't touch this function: I think it is because we use this only for adding temporarily config parameters. But I check with Orel.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+
+    def change_parameter(self, par_name, value):
+        """
+        Change the value of an existing parameter to the model. This function does not modify the \
```
Fixed it (one of the functions I moved only, but didn't touch).

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+            self._logger.error(msg)
+            raise InvalidModelParameter(msg)
+
+        type_of_par_name = locate(self.get_parameter_type(par_name))
+        if not isinstance(value, type_of_par_name):
+            self._logger.warning(
+                f"The type of the provided value ({value}, {type(value)}) "
+                f"is different from the type of {par_name} "
+                f"({self.get_parameter_type(par_name)}). "
```
@orelgueta - can you comment on this? This is reviewed code I've moved, but didn't change.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        if is_file:
+            self._is_exported_model_files_up_to_date = False
+
+    def change_parameter(self, par_name, value):
```
Again, this is moved code without changes. I am hesitating to touch this (given that this PR is already very big and a refactoring of all the functions would take much more time). Definitely all good comments!

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        self._parameters[par_name]["value"] = value
+
+        # In case parameter is a file, the model files will be outdated
+        if self.get_parameter_file_flag(par_name):
+            self._is_exported_model_files_up_to_date = False
```
OK. But in most cases we use the model as a whole and not individual parameters. So one change makes a model being outdated. But let's discuss your proposal again and see if we should update this accordingly in a future PR.

VictorBarbosaMartins replied:
Yes. Let's leave this comment open for a while, I will have a look if I find a quick solution on Monday. Once simtools is close to complete and we have all the necessary derivation functions for all parameters, we would like to know which ones are not updated and update only those, because the derivation of some parameters might take some time (e.g. the random angle tool).

VictorBarbosaMartins replied:
Ok, as we spoke in person, this is not necessary, as the workflows will automatically update all parameters related to a parameter changed.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+
+        self._is_config_file_up_to_date = False
+
+    def change_multiple_parameters(self, **kwargs):
+        """
+        Change the value of multiple existing parameters in the model. This function does not \
```
Honestly I don't see the difference - I think both terms are the same / very similar.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+
+        Raises
+        ------
+        InvalidModelParameter
```
Removed this part.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        shutil.copy(file_path, self.config_file_directory)
+
+    def export_model_files(self):
+        """Exports the model files into the config file directory."""
```
This is code moved from `TelescopeModel` and I suggest not to change it (or change it in a future update). If we want to do that, then I suggest to open an issue (there is a 'Reference new issue' button on the drop down menu).

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+        )
+
+    @property
+    def config_file_directory(self):
+        """
+        Directory for configure files. Configure, if necessary.
+
+        """
+        if self._config_file_directory is None:
+            self._set_config_file_directory_and_name()
+        return self._config_file_directory
+
+    @property
+    def config_file_path(self):
```
This is how I learned the expressions, and whoever above used the expressions followed this logic (I've added the properties in this case):

- path of a file is directory + file name
- directory of a file is directory only
- path of a directory is the directory (here the terms are the sames)

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+    def get_config_file(self, no_export=False):
+        """
+        Get the path of the config file for sim_telarray. The config file is produced if the file\
+        is not updated.
```
I think it should mean 'up to date' and not 'updated. Fixed this.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,561 @@
+            self.export_config_file()
+        return self.config_file_path
+
+    def get_derived_directory(self):
+        """
+        Get the path where all the files with derived values for are written to.
```
This is correct. We return a directory here.

VictorBarbosaMartins replied:
but the docstring is referring to a `path`. it should then be `directory` in the docstring.

GernotMaier commented on simtools/model/site_model.py:
```diff
@@ -0,0 +1,114 @@
+
+        """
+
+        # backwards compatibility to `corsika_parameters.yml` (temporary)
```
Added a TODO.

Agree the first line is a bit weird but required to allow for similar outputs as in the past. Good to have a TODO there now, so that we remember to remove it future.

GernotMaier commented on simtools/model/site_model.py:
```diff
@@ -0,0 +1,114 @@
+            Simtel site parameters as dict
+
+        """
+        return super().get_simtel_parameters(telescope_model=telescope_model, site_model=site_model)
```
I think this makes it clear that we call the base class function - given that there is a `get_simtel_parameters` function in both (and I don't know enough about python to see if this is needed or not).

GernotMaier commented on simtools/model/telescope_model.py:
```diff
@@ -205,386 +162,12 @@ def _process_line(words):
-
-        self._is_config_file_up_to_date = False
-
-    def remove_parameters(self, *args):
```
It is nowhere used.

GernotMaier commented on simtools/model/telescope_model.py:
```diff
@@ -866,3 +428,15 @@ def export_table_to_model_directory(self, file_name, table):
+            simtel parameters as dict
+
+        """
+        return super().get_simtel_parameters(telescope_model=telescope_model, site_model=site_model)
```
Same comment as above. I think it makes it clearer.

GernotMaier commented on simtools/schemas/data.metaschema.yml:
```diff
@@ -283,10 +283,6 @@ definitions:
         type: string
         description: |-
           "version of simulation software this parameter is used"
-      parameter_name:
-        type: string
-        description: |-
```
This is a duplication between `parameter_name` and `name`, which has been removed.

GernotMaier commented on simtools/layout/telescope_position.py:
```diff
@@ -93,41 +95,38 @@ def print_compact_format(
             if print_header:
-                print(headerstr)
-            print(telstr)
+                print(header_str)
+            print(tel_str)
```
Wait - this is for the `print` function and we actually want to print (without the clutter added by logging).

GernotMaier commented on simtools/layout/telescope_position.py:
```diff
@@ -321,6 +323,22 @@ def _get_reference_system_from(self):
 
+    def is_coordinate_system(self, crs_name):
+        """
+        Check if crs_name describes a coordinate system or auxiliary information.
```
`crs` == `coordinate reference system` which we abbreviate by `coordinate_system`. So adding `crs` would be a duplication.

GernotMaier commented on simtools/layout/telescope_position.py:
```diff
@@ -321,6 +323,22 @@ def _get_reference_system_from(self):
+        ----------
+        crs_name: str
+            Name of coordinate system.
+
```
Added.

GernotMaier commented on simtools/layout/telescope_position.py:
```diff
@@ -503,11 +526,55 @@ def convert_all(self, crs_local=None, crs_wgs84=None, crs_utm=None):
+
+        Returns
+        -------
+        astropy.units.m
```
Fixed.

GernotMaier commented on simtools/model/array_model.py:
```diff
@@ -1,9 +1,9 @@
-from simtools import db_handler
 from simtools.io_operations import io_handler
 from simtools.layout.array_layout import ArrayLayout
+from simtools.model.site_model import SiteModel
```
Fixed.

GernotMaier commented on simtools/layout/array_layout.py:
```diff
@@ -34,29 +35,31 @@ class ArrayLayout:
         telescope_list_file=None,
         telescope_list_metadata_file=None,
         validate=False,
+        array_layout_name=None,
```
Good point. I've simplified this now and removed array_layout_name from the main function init.

GernotMaier commented on simtools/layout/array_layout.py:
```diff
@@ -551,68 +355,102 @@ def _try_set_altitude(self, row, tel, table):
         try:
             tel.set_altitude(
                 self._altitude_from_corsika_z(
```
corsika_z is not the axis height but the height above corsika_observation_level. So using `z` is probably consistent.

VictorBarbosaMartins replied:
I was trying to think about using another name, but it is indeed hard to come up with a ambiguous name for it without using "corsika" and "z". let's then leave it like this (I would only include a comment with what you just wrote to clarify it)

GernotMaier commented on simtools/layout/array_layout.py:
```diff
@@ -551,68 +355,102 @@ def _try_set_altitude(self, row, tel, table):
             pass
 
-    def _load_telescope_list(self, table):
+    def initialize_array_layout(
```
Yes, fixed.

GernotMaier commented on simtools/layout/array_layout.py:
```diff
@@ -551,68 +355,102 @@ def _try_set_altitude(self, row, tel, table):
         """
-        Load list of telescope from an astropy table (support both QTable and Table)
+        Initialize the Layout array including site and telescope parameters.
+        Read array lit if telescope_list_file is given.
```
Typo - should be `list`. Fixed.

GernotMaier commented on simtools/layout/array_layout.py:
```diff
@@ -551,68 +355,102 @@ def _try_set_altitude(self, row, tel, table):
-    def initialize_array_layout_from_telescope_file(
-        self, telescope_list_file, telescope_list_metadata_file=None, validate=False
-    ):
+        return table
```
It is used (at this point in tests only), but I think it is also useful.

GernotMaier commented on simtools/layout/array_layout.py:
```diff
@@ -551,68 +355,102 @@ def _try_set_altitude(self, row, tel, table):
+                    ] = tel_model.get_parameter_value_with_unit(para)
+
+            for key, value in self._auxiliary_parameters[_telescope_model_name].items():
+                telescope.set_auxiliary_parameter(key, value)
```
Can you be more detailed and come up with a suggestion? I struggle seeing the issue. 

Note that a bit of the complication comes from the fact that we want to make the data base calls in TelescopeModel only for each model class, not for each telescope.

VictorBarbosaMartins replied:
Ok, your comment convinced me, let's leave it like this then.

GernotMaier commented on simtools/utils/general.py:
```diff
@@ -882,3 +912,29 @@ def get_value_unit_type(value):
+
+def get_value_as_quantity(value, unit):
+    """
+    Get a value as a Quantity with a given unit. If value is a Quantity, convert to unit.
```
The dostring should say `convert to the given unit.`. Fixed.

GernotMaier commented on simtools/utils/general.py:
```diff
@@ -882,3 +912,29 @@ def get_value_unit_type(value):
+        try:
+            value = value.to(unit)
+            return value
+        except u.UnitConversionError:
```
Fixed.

GernotMaier commented on simtools/visualization/legend_handlers.py:
```diff
@@ -2,10 +2,8 @@
-from simtools.io_operations.io_handler import IOHandler
-from simtools.utils import names
-from simtools.utils.names import hess, lst, magic, mst, sct, sst, veritas
+# from simtools.corsika.corsika_config import CorsikaConfig
+# from simtools.io_operations.io_handler import IOHandler
```
Agree. fixed.

GernotMaier commented on simtools/visualization/legend_handlers.py:
```diff
@@ -30,32 +28,48 @@
     "TelescopeHandler",
 ]
 
+# assume same size of telescope at both sides
+# (good assumption for plotting)
+lst = "LSTN"
+mst = "MSTN"
+sst = "SSTS"
+sct = "SCTS"
+hess = "HESS"
+magic = "MAGIC"
+veritas = "VERITAS"
+
```
LegendHandler is a mess and it costs me already hours in the context of this PR for a simple function to plot a legend and handle the marker size depending on the telescope type. The telescope types had been hardwired before in here and I agree that this is not a good solution.

One of the issues is that we cannot define a TelescopeModel here, as we get a circular import error.

This module needs a complete refactoring and I move it to issue #818

GernotMaier commented on simtools/visualization/legend_handlers.py:
```diff
@@ -30,32 +28,48 @@
+            "VERITAS": "grey",
+        }
+
+        # hardwired values; this is for plotting purposes only
+        self.radius_dict = {
+            "LSTN": 12.5,
+            "MSTN": 9.15,
+            "LSTS": 12.5,
+            "MSTS": 9.15,
+            "SCTS": 7.15,
+            "SSTS": 3.0,
+            "HESS": 6.0,
+            "MAGIC": 8.5,
+            "VERITAS": 6.0,
```
See my comment above. I think the complete module requires a refactoring, see #818.

GernotMaier commented on simtools/visualization/legend_handlers.py:
```diff
@@ -419,17 +433,21 @@ def legend_artist(legend, orig_handle, fontsize, handlebox):
 
 
 all_telescope_objects = {
-    lst: LSTObject,
-    mst: MSTObject,
+    "LSTN": LSTObject,
+    "LSTS": LSTObject,
+    "MSTN": MSTObject,
+    "MSTS": MSTObject,
     sct: SCTObject,
     sst: SSTObject,
     hess: HESSObject,
     magic: MAGICObject,
     veritas: VERITASObject,
 }
 all_telescope_handlers = {
-    lst: LSTHandler,
-    mst: MSTHandler,
+    "LSTN": LSTHandler,
+    "LSTS": LSTHandler,
+    "MSTN": MSTHandler,
+    "MSTS": MSTHandler,
```
We should move all this into #818.

GernotMaier commented on simtools/visualization/legend_handlers.py:
As described below, you need to have access to the `model_parameters` repository to run these applications. I've just tested it again and it works fine for me. Let me know if not for you.

GernotMaier commented on simtools/visualization/visualize.py:
```diff
@@ -14,7 +14,6 @@
 
 from simtools.utils import geometry as transf
 from simtools.utils import names
-from simtools.utils.names import mst, sct
```
Yes - and this should be part of the refactoring of the legend_handler class.

GernotMaier commented on tests/unit_tests/layout/test_array_layout.py:
```diff
@@ -600,89 +403,46 @@ def test_one_site(instance, test_file, manual_xx, manual_yy):
+def test_export_telescope_list_table(
+    db_config, telescope_north_test_file, telescope_north_utm_test_file
 ):
-    _ground_table = QTable.read(telescope_north_test_file, format="ascii.ecsv")
-    _ground_layout = ArrayLayout(mongo_db_config=db_config, site="North")
-    _ground_layout._load_telescope_list(_ground_table)
-    assert len(_ground_layout._telescope_list) == 13
-    assert _ground_layout._telescope_list[0].crs["ground"]["xx"]["value"] == pytest.approx(
-        _ground_table["position_x"][0].value
-    )
-    assert _ground_layout._telescope_list[0].crs["ground"]["yy"]["value"] == pytest.approx(
-        _ground_table["position_y"][0].value
-    )
-
-    _utm_table = QTable.read(telescope_north_utm_test_file, format="ascii.ecsv")
-    _utm_layout = ArrayLayout(mongo_db_config=db_config, site="North")
-    _utm_layout._load_telescope_list(_utm_table)
-    assert len(_utm_layout._telescope_list) == 13
-    assert _utm_layout._telescope_list[0].crs["utm"]["xx"]["value"] == pytest.approx(
-        _utm_table["utm_east"][0].value
-    )
-    assert _utm_layout._telescope_list[0].crs["utm"]["yy"]["value"] == pytest.approx(
-        _utm_table["utm_north"][0].value
-    )
-
-    _mercator_table = QTable.read(telescope_north_mercator_test_file, format="ascii.ecsv")
-    _mercator_layout = ArrayLayout(mongo_db_config=db_config, site="North")
-    _mercator_layout._load_telescope_list(_mercator_table)
-    assert len(_mercator_layout._telescope_list) == 13
-    assert _mercator_layout._telescope_list[0].crs["mercator"]["xx"]["value"] == pytest.approx(
-        _mercator_table["latitude"][0].value
-    )
-    assert _mercator_layout._telescope_list[0].crs["mercator"]["yy"]["value"] == pytest.approx(
-        _mercator_table["longitude"][0].value
```
The removed tests essentially duplicated unit tests from astropy and tested that astropy.tables can read the data and metadata from an ecsv file. I don't think this is necessary.

(note also that `_load_telescope_list` does not exist anymore. The functionality is in `initialize_array_layout`).

GernotMaier commented on tests/unit_tests/model/test_site_model.py:
```diff
@@ -0,0 +1,28 @@
+def test_site_model(db_config):
+    _south = SiteModel(site="South", mongo_db_config=db_config, label="testing-sitemodel")
+
+    assert isinstance(_south.get_reference_point(), dict)
```
Discussed this further above. I would like to have the number of fixed model values in the unit test reduced, as these values might change. Also not convinced that this is necessary to test the function.

GernotMaier commented on tests/unit_tests/model/test_model_parameter.py:
```diff
@@ -0,0 +1,192 @@
+import pytest
+from astropy import units as u
+
+from simtools.model.model_parameter import InvalidModelParameter
```
I can see the asymmetry in the naming, but this is not the model itself, but the base class which mostly provides the functions required to retrieve parameters from the DB or manipulate them. Is is therefore more like the `model_utils` module than on of the telescope/site models. My preference would to keep it, but we obviously should come to an agreement on this).

VictorBarbosaMartins replied:
I understand your explanation above, but this name causes confusion at least to me. Based on this name, I thought the class would deal with a single parameter, therefore my surprise when I found this is dealing with all the parameters, which means, it is dealing in fact with the simulation model itself. What about renaming it to `Parameters` ?

GernotMaier commented on tests/unit_tests/test_camera_efficiency.py:
```diff
@@ -66,27 +72,30 @@ def test_validate_telescope_model(simtel_path):
     assert (
         camera_efficiency_lst._file_log.name
-        == "camera-efficiency-North-LST-1-za020deg_azm000deg_validate_camera_efficiency.log"
+        == "camera-efficiency-North-LSTN-01-za020deg_azm000deg_test-telescope-model-lst.log"
```
As discussed in this person, this is beyond the current PR (and outlined in #804).

GernotMaier commented on tests/unit_tests/test_ray_tracing.py:
```diff
@@ -137,6 +137,7 @@ def test_ray_tracing_invalid_telescope_model(simtel_path, io_handler, caplog):
         assert "Invalid TelescopeModel" in caplog.text
 
 
+@pytest.mark.skip(reason="Files from test-data DB to be move to ./tests/resources directory")
```
As discussed in this person, this is beyond the current PR (and outlined in #804).

GernotMaier commented on tests/unit_tests/test_simulator.py:
```diff
@@ -88,7 +88,7 @@ def test_guess_run_from_file(array_simulator):
     assert array_simulator._guess_run_from_file("abc-ran12345_bla_ble") == 1
 
-    # TODO add test for 'Error creating output directory'
+    # add test for 'Error creating output directory'
```
To `TODO` has been removed as it is not necessary to mark missing tests in the code (this is done by the coverage tools). I've kept the comments as there is an additional line indicating what the challenge is. If you prefer, we can remove it - but we loose this information then.

GernotMaier commented on tests/unit_tests/visualization/test_legend_handlers.py:
```diff
@@ -40,8 +40,8 @@ def test_handlers(io_handler):
-    for step, tel_type in enumerate(names.all_telescope_class_names):
-        assert tel_handler.radius_dict[tel_type] == pytest.approx(radius_dict[step], 1.0e-3)
-        assert tel_handler.colors_dict[tel_type] == colors[step]
+    assert len(tel_handler.colors_dict) == len(
+        names.get_list_of_telescope_types(
+            array_element_class="telescope", site=None, observatory=None
+        )
+    )
```
I expect that legendHandler will be refactored (#818) and I refer this to there.

GernotMaier commented on tests/unit_tests/utils/test_names.py:
```diff
@@ -636,3 +457,23 @@ def test_camera_efficiency_log_file_name():
+    assert names.get_simtel_name_from_parameter_name("telescope_axis_height") is None
+
+    assert (
+        names.get_simtel_name_from_parameter_name("corsika_observation_level", site_model=False)
```
This is a different naming (sorry) - here it is about how we have saved it in the database. This function ensures backwards compatibility with the existing models in the database. Naming can be so confusing!!

GernotMaier commented:
@VictorBarbosaMartins  - thanks again for the careful review and the useful comments!

I've made a lot of improvements and changes following your comments, please have another look.

I think we have discussed most of the issues in person:

- most of the code in `model_parameter` has been moved there, but not changed. Although it has been reviewed, there is obviously always room for improvement. But as discussed, I don't see this as something to be done in this PR, as we would significantly increase the already to complex updates
- naming is still a mess and not ideal. I think this PR makes a good step forward to improve it and as soon as the DB is updated with all parameters and the right naming we can simplify some of the code. I've tried to move (almost) all of the naming complexity into naming.py, hopefully this will make future changes easier
- some of your points have been moved to issues to keep this PR in a reasonable size (especially legend_handler, which is a module I am suggesting for refactoring for a long time)

We should discuss the additional splitting of model parameters into individual parameters. I've carefully looked through the ModelParameter class, but your suggestion is not obvious to me. Best to discuss this in person.

I think I've addressed most of your points, please let me know if I've missed something.

VictorBarbosaMartins reviewed (approved):
Hi @GernotMaier.
I am approving the PR but left some conversations open with some minor points. Please feel free to decide what to do in these regards. Thanks a lot for this major change, it will hopefully help us deal with the parameters and facilitate the further development of the code.

VictorBarbosaMartins commented on simtools/utils/general.py:
```diff
@@ -913,6 +916,15 @@ def get_value_unit_type(value):
         base_value = value
         base_type = extract_type_of_value(base_value)
 
+    if unit_str is not None:
+        if base_unit is not None:
+            try:
+                base_value = base_value * u.Unit(base_unit).to(u.Unit(unit_str))
+            except u.UnitConversionError:
+                _logger.error(f"Cannot convert {base_unit} to {unit_str}.")
+                raise
+        base_unit = unit_str
+
```
These changes made me realize that the function returns `value` and `unit`, and `value` can even a `Quantity`. This is confusing. `value` shouldn't be `Quantity`. If you need the quantity instance, then I would call it quantity instead of value.

VictorBarbosaMartins commented on simtools/applications/db_development_tools/add_new_parameter_from_json_file_to_db.py:
```diff
@@ -0,0 +1,61 @@
+    This application is used to add a new parameter to the sites collection in the DB
+    using a json file as input.
+
+    This application should not be used by anyone but expert users and not often.
+    Therefore, no additional documentation about this applications will be given.
```
Experts and developers also need documentation (and tests). I thought with the PR approved we wouldn't need an additional application.

VictorBarbosaMartins commented on simtools/applications/db_development_tools/add_new_parameter_from_json_file_to_db.py:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    for version, value in value_and_version.items():
+        db.add_new_parameter(
+            db_name=db.DB_CTA_SIMULATION_MODEL,
+            telescope=par_dict["instrument"],
+            parameter=par_dict["parameter"],
+            version=version,
+            value=value,
+            site=par_dict["site"],
+            type=par_dict["type"],
+            collection_name=args_dict["db_collection"],
+            applicable=par_dict["applicable"],
+            file=par_dict["file"],
+            unit=par_dict["unit"] if "unit" in par_dict and par_dict["unit"] is not None else None,
+            file_prefix="./",
```
houldn't we check whether all these keys exist before using it? I would rather use a `try` and `except` here

orelgueta commented on simtools/utils/names.py:
```diff
@@ -257,322 +249,202 @@ def _validate_name(name, all_names):
-        _telescope_name += "-" + tel_id
-    return _telescope_name
+        _tel_type, _tel_id = name.split("-")
+    except ValueError as exc:
```
When would this return a ValueError?

GernotMaier replied:
e.g. `_tel_type, _tel_id = "LSTN".split("-")` , as it has only one return parameter ("LSTN").

orelgueta reviewed (commented):
Very nice!

I made just a few comments and only one of them worries me a bit (about the Structure/Camera readout from the DB).

GernotMaier commented on simtools/camera_efficiency.py:
```diff
@@ -452,7 +451,7 @@ def calc_nsb_rate(self):
         nsb_rate_ref_conditions = (
             nsb_rate_provided_spectrum
-            * self._telescope_model.reference_data["nsb_reference_value"]["Value"]
+            * self._telescope_model.reference_data["nsb_reference_value"]["value"]
```
OK, I've added this by:

- moving reference parameter from `telescope_model` to `model_parameter (as e.g., the `SiteModel` might also want to get access to reference data
- added a `get_reference_data_value` method as suggested
- added unit tests.

GernotMaier commented on simtools/data_model/data_reader.py:
```diff
@@ -120,10 +120,11 @@ def read_value_from_file(file_name, schema_file=None, validate=False):
+    _value = data.get("value")
+    if _value is None:
         return None
+    _units = data.get("unit")
```
Fixed.

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -343,48 +257,21 @@ def _get_model_parameters_mongo_db(
         for _tel in _which_tel_labels:
             self._logger.debug(f"Getting {_tel} parameters from MongoDB")
 
-            # If tel is a structure, only applicable pars will be collected, always.
```
Not sure I understand the implication.

We have decided to get rid of Structure and Camera and use only the full telescope model name: `LSTN-01`, `LSTN-design`, etc.

orelgueta replied:
That's the change of strategy I meant, so hopefully all is well. We can discuss it in a short call, it would be quicker to explain the implication.

GernotMaier commented:
![Screenshot 2024-03-06 at 16 00 35](https://github.com/gammasim/simtools/assets/14001688/6f111ab0-af06-4201-9b90-681068744c66)

Github does not allow me to reply on that: 'fixed'.

GernotMaier commented on simtools/layout/array_layout.py:
```diff
@@ -551,68 +353,102 @@ def _try_set_altitude(self, row, tel, table):
+        if names.get_class_from_telescope_name(telescope.name) == "telescope":
+            _telescope_model_name = self.db.get_telescope_db_name(telescope.name)
+            self._logger.info(
+                f"Reading auxiliary telescope parameters for {telescope.name}"
+                f" (telescope model {_telescope_model_name})"
+            )
+            if _telescope_model_name not in self._auxiliary_parameters:
+                tel_model = TelescopeModel(
+                    site=self.site,
+                    telescope_model_name=_telescope_model_name,
+                    model_version=self.model_version,
+                    mongo_db_config=self.mongo_db_config,
+                    label=self.label,
+                )
+                self._auxiliary_parameters[_telescope_model_name] = {}
+                for para in ("telescope_axis_height", "telescope_sphere_radius"):
+                    self._auxiliary_parameters[_telescope_model_name][para] = (
+                        tel_model.get_parameter_value_with_unit(para)
+                    )
+
+            for key, value in self._auxiliary_parameters[_telescope_model_name].items():
+                telescope.set_auxiliary_parameter(key, value)
```
Yes, and they are read in this part using db_handler in TelescopeModel (although db_handler internally routes it to the db_repo_handler and reads these parameters, until added to the db, from the repository).

This solution reads the telescope model once and saves the parameters for all telescopes. I guess this should be fine, or is there something wrong in this implementation?

orelgueta replied:
Not the implementation, I am just wondering about having these parameters as auxiliary ones instead of adding them as part of the "nominal" model like I thought we planned. If you prefer to keep it like this, no problem.

GernotMaier commented on simtools/model/mirrors.py:
```diff
@@ -77,8 +79,8 @@ def _read_mirror_list_from_ecsv(self):
                 self.mirror_diameter = u.Quantity(
-                    self.parameters["mirror_panel_diameter"]["Value"],
-                    self.parameters["mirror_panel_diameter"]["units"],
+                    self.parameters["mirror_panel_diameter"]["value"],
```
Not in the implementation we have: we pass on a dict from TelescopeModel to Mirrors, so here we cannot use the model_parameter functions. 

In general (as mentioned a couple of times), we need to rethink the organisation of the model parameters from the high level (Array/Site) down to mirrors and pixels.

orelgueta replied:
OK, it can be dealt with later then.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,556 @@
+        _value = self.get_parameter_value(None, _parameter)
+        try:
+            _units = _parameter.get("unit") or _parameter.get("units")
+            return float(_value) * u.Unit(_units)
```
This is an astropy quantity, which is always floating point (I don't think we have integer values with units?)

orelgueta replied:
OK!

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,556 @@
+        _parameter = self.get_parameter_dict(par_name)
+        _value = self.get_parameter_value(None, _parameter)
+        try:
+            _units = _parameter.get("unit") or _parameter.get("units")
```
We can remove it now - that is a leftover from development where I had things coming in with the old and new naming. I've removed it.

GernotMaier commented on simtools/model/model_parameter.py:
```diff
@@ -0,0 +1,556 @@
+        """
+        parameter_dict = self.get_parameter_dict(par_name)
+        try:
+            return parameter_dict.get("type") or parameter_dict.get("Type")
```
Removed it.

GernotMaier commented:
![Screenshot 2024-03-06 at 16 21 20](https://github.com/gammasim/simtools/assets/14001688/27d936ef-e970-4ba6-8db7-df85269f4424)

Propagated this to issue #828 (and again I cannot answer to this comment)

GernotMaier commented on simtools/model/site_model.py:
```diff
@@ -0,0 +1,114 @@
+
+        # backwards compatibility to `corsika_parameters.yml` (temporary TODO)
+        if config_file_style:
+            _atmosphere_id = 26 if self.site == "North" else 36
```
Issue #829

GernotMaier commented on simtools/utils/names.py:
```diff
@@ -19,59 +17,53 @@
+    "SSTS": {"site": "South", "observatory": "CTAO", "class": "telescope"},
+    "SCTS": {"site": "South", "observatory": "CTAO", "class": "telescope"},
+    # calibration devices
+    "ILLN": {"site": "North", "observatory": "CTAO", "class": "calibration"},
```
You mean what `ILLN` North means? I think I remember.

I anyway plan to remove all those lines and move them into the json schema (some parts are already duplicated, e.g., the list of telescopes) - names.py will read then the schema files to get the naming dicts.

orelgueta replied:
Yes, ILLN, RLDN, STPN, etc. I don't remember what all of them means and I think it would be useful to have the non-abbreviated names as well. Also if it's in the schema files it would be useful.

GernotMaier commented on simtools/utils/names.py:
```diff
@@ -88,13 +80,14 @@
     "2019-12-30": [""],
     "2020-02-26": [""],
     "2020-06-28": ["prod5"],
+    "2024-02-01": [""],
```
I don't remember - changed it to Feb 5th.

GernotMaier commented:
@orel - thanks for the review!! I've addressed all of your comments.

I will need your help on the `db_handler:_get_model_parameters_mongo_db` for the selection of the applicable parameters. What is wrong with the implementation now? Is it inefficient, as we get the full telescope model instead of camera or structure only? Or is there another issue which I don't see?
